internal/proxy/multi_app_conn.go

```diff
@@ -147,26 +148,63 @@ func (app *multiAppConn) killTMOnClientError() {
 		}
 	}
 
+	type op struct {
+		connClient stoppableClient
+		name       string
+	}
+
+	wg := &sync.WaitGroup{}
+	for _, client := range []op{
+		{
+			connClient: app.consensusConnClient,
+			name:       connConsensus,
+		},
+		{
+			connClient: app.mempoolConnClient,
+			name:       connMempool,
+		},
+		{
+			connClient: app.queryConnClient,
+			name:       connQuery,
+		},
+		{
+			connClient: app.snapshotConnClient,
+			name:       connSnapshot,
+		},
+	} {
+		wg.Add(1)
+		go func(client op) {
+			defer wg.Done()
+
+			select {
+			case <-ctx.Done():
+			case <-signalWait(client.connClient.Wait):
+				if err := client.connClient.Error(); err != nil {
+					killFn(client.name, err, app.Logger)
+				}
+			}
+		}(client)
+	}
+
+	// convert wg into a signal channel so we can cleanup if no
+	// one is waiting
 	select {
-	case <-app.consensusConnClient.Quit():
-		if err := app.consensusConnClient.Error(); err != nil {
-			killFn(connConsensus, err, app.Logger)
-		}
-	case <-app.mempoolConnClient.Quit():
-		if err := app.mempoolConnClient.Error(); err != nil {
-			killFn(connMempool, err, app.Logger)
-		}
-	case <-app.queryConnClient.Quit():
-		if err := app.queryConnClient.Error(); err != nil {
-			killFn(connQuery, err, app.Logger)
-		}
-	case <-app.snapshotConnClient.Quit():
-		if err := app.snapshotConnClient.Error(); err != nil {
-			killFn(connSnapshot, err, app.Logger)
-		}
```

Comment:
Am I reading this wrong or was this select statement only waiting for one of the four to stop rather than all four?

Reply:
this is in a method called `killTMOnClientError` so I think you've identified a behavior change, though I don't know exactly what we should do.

Previously if any ABCI client returned an error, this function would kill _tendermint_ (via SIGTERM'ing our own PID), and my change means that we may attempt to kill tendermint more times, which clearly won't do anything...

Anyway, I think I can change this to just start N go routines and avoid the wait group.

Reply:
Does it kill Tendermint multiple times or just a certain connection. Is Tendermint really just a Horcrux?

Reply:
Originally, it only killed it once. My first change killed it multiple times in a very careful way, and my second change is somewhat less careful, but equally good. 

There's no harm in issuing more than one SIGTERM, but I don't think we need to wait to send all SIGTERMs in a thread that doesn't return anything outside of itself.

Reply:
I'm maybe confused why we issue a SIGTERM for our PID instead of just `os.Exit`ing. That aside, it looks like the behavior here has been updated to roughly match the old behavior: only one of the clients needs to error for us to kill tendermint.

Reply:
> I'm maybe confused why we issue a SIGTERM for our PID instead of just os.Exiting. 

I mean I think both are probably bad. SIGTERM maybe lets us flush something to the database before exiting, because it means some of the cleanup/close logic in the node object would get called.

> That aside, it looks like the behavior here has been updated to roughly match the old behavior: only one of the clients needs to error for us to kill tendermint.

Not that it particularly matters, but I think even the intermittent form had the same semantics: the first client to return, would issue a sigterm, and, eventually, that process would die, though this goroutine would be waiting on the waitgroup when it exited, rather than having this particular thread return before the process exited, which doesn't actually change anything. In any case it's moot.

Reply:
Oof, I see, so we used SIGTERM to communicate that the root of the application should gracefully cleanup its resources. Gross.